tempodb/backend/gcs/gcs.go

```diff
@@ -319,3 +300,47 @@ func (rw *readerWriter) readRange(ctx context.Context, name string, offset int64
 		totalBytes += byteCount
 	}
 }
+
+func createBucket(ctx context.Context, cfg *Config, hedge bool) (*storage.BucketHandle, error) {
+	// start with default transport
+	customTransport := http.DefaultTransport.(*http.Transport).Clone()
+
+	// add google auth
+	transportOptions := []option.ClientOption{
+		option.WithScopes(storage.ScopeReadWrite),
+	}
+	if cfg.Insecure {
+		transportOptions = append(transportOptions, option.WithoutAuthentication())
+		customTransport.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+	}
+	transport, err := google_http.NewTransport(ctx, customTransport, transportOptions...)
+	if err != nil {
+		return nil, errors.Wrap(err, "creating google http transport")
+	}
+
+	// add instrumentation
+	transport = newInstrumentedTransport(transport)
```

Comment:
Thinking about this some - should we move instrumentation to after hedging in the middleware pipeline?

Reply:
We could get a higher rate of errors than we actually have because of all the retried and cancelled requests. I would instrument the hedging itself maybe.

Reply:
this should instrument both requests independently. i think this is what we want. it shows us our "true" gcs requests.

Reply:
OK. Do correct me if I'm wrong, but what I'm concerned about is that at the moment we don't have a lot of visibility of the number of hedged requests we're making in our deployment. As an example, if our gcs p99 increases to say 700ms, maybe due to some bloom filter/index changes we make in the future, the current instrumentation may "hide" it -

```
without hedged request:
gcs get request -> 700ms
```

```
with hedged:
gcs get request first try -> 500ms -> timed out -> second try -> 50ms
```

While this is exactly the behaviour that we want, this will cause our gcs p99 to flatline at 500ms. we wouldn't know what value we should set for the `hedge_requests_at` config parameter. A retry metric might we useful to track in that case

Reply:
This is not entirely correct: in my understanding the first request will not time out when the hedged request starts, it will not be cancelled unless the second request completes. So it would be like this:

```
1st request starts ->                 ...               -> 1st request is cancelled because 2nd completed
                   -> after 500 ms 2nd request starts -> 2nd request completes after 50 ms
```

In this case the first request had a latency of 550 ms.

If the second request is slower, the first one will not be cancelled:

```
1st request starts ->                 ...             -> 1st request completes after 700 ms
                   -> after 500 ms 2nd request starts   -> 2nd request is cancelled
```

So the duration of the first request should still show us the total duration of the operation.

---

Maybe it's interesting to track how often the hedged request wins? This can help us when tuning the system: if the hedged requests is never faster, we shouldn't spend resources launching it.

Reply:
The maintainers would need to accept a PR where hedgedghttp takes a prometheus registry and then we do some internal metrics. Personally I'd prefer moving forward as is. Although I agree there will be a small lack of visibility here.

Reply:
Feel free to raise a feature request. I think something simple like counters can be added. However I would like to keep library without dependencies (especially Prometheus :)

Maybe it possible to solve with a custom `RoundTripper`, which can track distinguish 1st and any other consecutive request (probably by url or even `*req` address) and update proper metrics?

(also my 5c: agree with @joe-elliott it's a good feature for the future)